include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/sort/common.hlsl"
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+NBL_CONSTEXPR uint32_t BucketsPerThread = ceil((float) BucketCount / WorkgroupSize);
+
+groupshared uint32_t prefixScratch[BucketCount];
+
+struct ScratchProxy
+{
+    uint32_t get(const uint32_t ix)
+    {
+        return prefixScratch[ix];
+    }
+    void set(const uint32_t ix, const uint32_t value)
+    {
+        prefixScratch[ix] = value;
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        nbl::hlsl::glsl::barrier();
+    }
+};
+
+static ScratchProxy arithmeticAccessor;
+
+groupshared uint32_t sdata[BucketCount];
+
+template<typename KeyAccessor, typename ValueAccessor, typename ScratchAccessor>
+struct counting
+{
+    void histogram(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ScratchAccessor) scratch, const nbl::hlsl::sort::CountingPushData data)
+    {
+        uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+        [unroll]
+        for (int i = 0; i < BucketsPerThread; i++)
+            sdata[BucketsPerThread * tid + i] = 0;
+        uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize) * data.elementsPerWT;
+
+        nbl::hlsl::glsl::barrier();
+
+        for (int i = 0; i < data.elementsPerWT; i++)
+        {
+            int j = index + i * WorkgroupSize + tid;
+            if (j >= data.dataElementCount)
+                break;
+            uint32_t k = key.get(j);
+            nbl::hlsl::glsl::atomicAdd(sdata[k - data.minimum], (uint32_t) 1);
+        }
+
+        nbl::hlsl::glsl::barrier();
+
+        uint32_t sum = 0;
+        uint32_t scan_sum = 0;
+
+        for (int i = 0; i < BucketsPerThread; i++)
+        {
+            sum = nbl::hlsl::workgroup::exclusive_scan < nbl::hlsl::plus < uint32_t >, WorkgroupSize > ::
+            template __call <ScratchProxy>
+            (sdata[WorkgroupSize * i + tid], arithmeticAccessor);
+
+            arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
+
+            nbl::hlsl::glsl::atomicAdd(scratch.get_ptr(WorkgroupSize * i + tid), sum);
+            if ((tid == WorkgroupSize - 1) && i > 0)
+                nbl::hlsl::glsl::atomicAdd(scratch.get_ptr(WorkgroupSize * i), scan_sum);
```

Comment:
this doesn't need to be sandwidched between barriers, these can happen whenever.

Also try to precompute expressions like `WorkgroupSize*i+tid` or `WorkgroupSize-1` not because of speed, but because of readability

Reply:
btw a way to optimize this would be not to do `atomicAddIntoGlobal` but put the results back into `sharedScratch` and copy back this stuff later in another loop where you slightly rotate/stagger what index you start at depending on the workgroup ID, I can explain this later if you want

Reply:
Also don't use `get_ptr()` on your accessor and `glsl::atomicAdd`, reply on the accessor to have an `atomicAdd(index,value)` method instead, just how our workgroup ballot code does (see ex 23 accessors)

Reply:
I haven't done the optimization you suggested yet but the rest are done

Reply:
Optimization done as well